File: testdata/python/vulnerable.py
"""Intentionally insecure Python code for testing purposes."""
import hashlib
import os
import pickle
import subprocess


# CWE-798: Hardcoded credential
HARDCODED_PASSWORD = "super_secret_password"


def sql_injection(conn, user_input):
    """CWE-89: SQL injection via string formatting."""
    cursor = conn.cursor()
    query = "SELECT * FROM users WHERE name = '%s'" % user_input
    cursor.execute(query)
    return cursor.fetchall()


def command_injection(user_input):
    """CWE-78: OS command injection."""
    return subprocess.check_output("ls " + user_input, shell=True)


def weak_hash(password):
    """CWE-327: Use of broken MD5 hash algorithm."""
    return hashlib.md5(password.encode()).hexdigest()


def path_traversal(user_path):
    """CWE-22: Path traversal via unsanitized user input."""
    base = "/var/data/"
    return open(base + user_path).read()


def insecure_deserialization(data):
    """CWE-502: Deserialization of untrusted data."""
    return pickle.loads(data)


def xml_injection(user_input):
    """CWE-611: XML External Entity via lxml without safe parsing."""
    import lxml.etree as ET
    parser = ET.XMLParser(resolve_entities=True)
    return ET.fromstring(user_input, parser)


def no_docstring_function(x, y):
    return x + y

File: testdata/go/vulnerable.go
// Package vulnerable contains intentionally insecure Go code for testing purposes.
package vulnerable

import (
	"crypto/md5"
	"database/sql"
	"fmt"
	"math/rand"
	"net/http"
	"os"
	"os/exec"
)

// HardcodedSecret stores a plaintext credential (CWE-798).
const HardcodedSecret = "password123"

// WeakHash uses MD5 for password hashing (CWE-327).
func WeakHash(input string) string {
	h := md5.New()
	h.Write([]byte(input))
	return fmt.Sprintf("%x", h.Sum(nil))
}

// SQLInjection builds a query via string concatenation (CWE-89).
func SQLInjection(db *sql.DB, userInput string) (*sql.Rows, error) {
	query := "SELECT * FROM users WHERE name = '" + userInput + "'"
	return db.Query(query)
}

// CommandInjection passes user input directly to a shell command (CWE-78).
func CommandInjection(userInput string) ([]byte, error) {
	cmd := exec.Command("sh", "-c", "ls "+userInput)
	return cmd.Output()
}

// InsecureRandom uses a non-cryptographic RNG for token generation (CWE-338).
func InsecureRandom() int {
	return rand.Int()
}

// PathTraversal uses user-supplied path without sanitization (CWE-22).
func PathTraversal(userPath string) ([]byte, error) {
	return os.ReadFile("/var/data/" + userPath)
}

// OpenRedirect redirects to a user-controlled URL without validation (CWE-601).
func OpenRedirect(w http.ResponseWriter, r *http.Request) {
	redirectURL := r.URL.Query().Get("url")
	http.Redirect(w, r, redirectURL, http.StatusFound)
}
